### DISCUSSION TITLE
Equivalent circuit model

### DISCUSSION THREAD

**Discussion by balaji3112** [pybamm-team/liionpack#237] [Q&A]
I wish to run liionpack simulation using thevenin model, but I'm facing lot of issues while trying to use thevenin model. for example the following key mismatch error. Does liionpack can only simulate SPM models or is it possible to simulate other models also?

    raise KeyError(f"'{key}' not found. Best matches are {best_matches}")
KeyError: "'Cell capacity [A.h]' not found. Best matches are ['Nominal cell capacity [A.h]']"

Process finished with exit code 1

**Comment by MustafaDragonborne** [comment]
I found a very similar problem when trying to simulate the thevenin model. I managed to define the thevenin model and the ECM parameter values but ran into a key error for the open circuit voltage. After a bit of debugging, I realized that in line 329 of solvers.py, the self.variable_names is defined as a list in which one of the variable names is "Measured battery open circuit voltage [V]". This is a variable name of the electrochemical models and I had to change it to "Open circuit voltage [V]" to have that running for me. So you might have to also do something similar for "Cell capacity [A.h]". But now I struggle with yet another error.

miniconda3/envs/testenv2/lib/python3.9/site-packages/liionpack/netlist_utils.py:524: RuntimeWarning: divide by zero encountered in divide
  g[R_map] = 1 / value[R_map]  # conductance = 1 / R

The complete error is as follows:

tepping simulation:  30%|██▉       | 11463/38281 [03:44<08:46, 50.95it/s]/home/<user>/miniconda3/envs/testenv2/lib/python3.9/site-packages/liionpack/netlist_utils.py:524: RuntimeWarning: divide by zero encountered in divide
  g[R_map] = 1 / value[R_map]  # conductance = 1 / R
/home/<user>/miniconda3/envs/testenv2/lib/python3.9/site-packages/scipy/sparse/linalg/_dsolve/linsolve.py:276: MatrixRankWarning: Matrix is exactly singular
  warn("Matrix is exactly singular", MatrixRankWarning)
psetup failed: .../casadi/interfaces/sundials/cvodes_interface.cpp:650: 'jacF' calculation failed
psetup failed: .../casadi/interfaces/sundials/cvodes_interface.cpp:650: 'jacF' calculation failed
psetup failed: .../casadi/interfaces/sundials/cvodes_interface.cpp:650: 'jacF' calculation failed
psetup failed: .../casadi/interfaces/sundials/cvodes_interface.cpp:650: 'jacF' calculation failed
psetup failed: .../casadi/interfaces/sundials/cvodes_interface.cpp:650: 'jacF' calculation failed
psetup failed: .../casadi/interfaces/sundials/cvodes_interface.cpp:650: 'jacF' calculation failed
psetup failed: .../casadi/interfaces/sundials/cvodes_interface.cpp:650: 'jacF' calculation failed
psetup failed: .../casadi/interfaces/sundials/cvodes_interface.cpp:650: 'jacF' calculation failed
psetup failed: .../casadi/interfaces/sundials/cvodes_interface.cpp:650: 'jacF' calculation failed
psetup failed: .../casadi/interfaces/sundials/cvodes_interface.cpp:650: 'jacF' calculation failed
At t = 0 and h = 1.3093e-11, the corrector convergence test failed repeatedly or with |h| = hmin.
Stepping simulation:  30%|██▉       | 11463/38281 [03:44<08:46, 50.96it/s]
Traceback (most recent call last):
  File "/home/<user>/miniconda3/envs/testenv2/lib/python3.9/site-packages/liionpack/solvers.py", line 375, in _step_solve_step
    vlims_ok = self._step(step, external_variables)
  File "/home/<user>/miniconda3/envs/testenv2/lib/python3.9/site-packages/liionpack/solvers.py", line 450, in _step
    self.step_actors()
  File "/home/<user>/miniconda3/envs/testenv2/lib/python3.9/site-packages/liionpack/solvers.py", line 648, in step_actors
    events = self.actors[0].step(self.build_inputs()[0])
  File "/home/<user>/miniconda3/envs/testenv2/lib/python3.9/site-packages/liionpack/solvers.py", line 80, in step
    self.step_solutions, self.var_eval, self.events_eval = self.step_fn(
  File "/home/<user>/miniconda3/envs/testenv2/lib/python3.9/site-packages/liionpack/solver_utils.py", line 103, in _serial_step
    casadi_sol = integrator(x0=x0, z0=z0, p=inputs)
  File "/home/<user>/miniconda3/envs/testenv2/lib/python3.9/site-packages/casadi/casadi.py", line 7772, in __call__
    return self.call(kwargs)
  File "/home/<user>/miniconda3/envs/testenv2/lib/python3.9/site-packages/casadi/casadi.py", line 6945, in call
    return _casadi.Function_call(self, *args)
RuntimeError: .../casadi/interfaces/sundials/cvodes_interface.cpp:392: CVode returned "CV_CONV_FAILURE". Consult CVODES documentation.


Any solutions to this problem is highly appreciated.

**Reply by balaji3112** [reply]
Thanks for the info, I will update the parameter keys.

**Reply by TomTranter** [reply]
@MustafaDragonborne looks like something going wrong with internal resistance calculation. Can you run the simulation up to just before you get the failure and check all the battery states of charge. Maybe one is getting close to 0 or full charge and erroring. Hard to help more without seeing script and results myself

**Reply by TomTranter** [reply]
Also, looks like the "Measured..." variable has been renamed in the PyBaMM develop branch for the next release. I have another PR open that should address this

**Reply by MustafaDragonborne** [reply]
So I managed to solve the problem by doubling the current of my drive cycle since I was simulating two cells in parallel but I'm not sure why doubling the current works at all. Like why should a lower current value cause an error? I started out at an initial SoC of around 0.5 and the likelihood of the SoC values going above or below its limits should be even less.

**Reply by MustafaDragonborne** [reply]
I also couldn't find the SoC as one of the output variables which I could then plot. Maybe a fix for this would also be very nice.

**Reply by TomTranter** [reply]
Did you change the busbar resistances from the defaults by very much. If these are too high compared to the internal resistance of the cell then you can get unstable results and one cell being dominant in charge and discharge. If you run higher current then you get higher internal resistance. That's my guess without seeing your code

**Reply by TomTranter** [reply]
You're right that SoC is not a variable you can output directly. To be honest I'm not entirely sure why not but it's a bit subjective to define as you can normalize within a voltage range of your choice. We do allow you to specify it to initialize the simulations though... so assumptions are being made already. If you look at particle concentrations this is the SoC effectively

**Reply by MustafaDragonborne** [reply]
> Did you change the busbar resistances from the defaults by very much. If these are too high compared to the internal resistance of the cell then you can get unstable results and one cell being dominant in charge and discharge. If you run higher current then you get higher internal resistance. That's my guess without seeing your code

No I used all the default values for the time being.

**Reply by dice-kay** [reply]
I had a very similar problem with PyBaMM==23.3 and liionpack==0.3.4 but with PyBaMM==23.2 and liionpack==0.3.3 worked.

**Comment by MustafaDragonborne** [comment]
@balaji3112 are you sure you have defined your ParameterValues object properly? They will not have the same parameter values as that of the electrochemical model parameter sets.
